pkg/client/testing/core/fixture.go

```diff
@@ -49,201 +59,310 @@ type ObjectScheme interface {
-// as a runtime.Object if Status == Success).  If multiple PodLists are provided, they
-// will be returned in order by the Kind call, and the last PodList will be reused for
-// subsequent calls.
-func NewObjects(scheme ObjectScheme, decoder runtime.Decoder) ObjectRetriever {
-	return objects{
-		types:   make(map[string][]runtime.Object),
-		last:    make(map[string]int),
+var _ ObjectTracker = &tracker{}
+
+// NewObjectTracker returns an ObjectTracker that can be used to keep track
+// of objects for the fake clientset. Mostly useful for unit tests.
+func NewObjectTracker(scheme ObjectScheme, decoder runtime.Decoder) ObjectTracker {
+	return &tracker{
 		scheme:  scheme,
 		decoder: decoder,
+		objects: make(map[unversioned.GroupVersionKind][]runtime.Object),
 	}
 }
 
-func (o objects) Kind(kind unversioned.GroupVersionKind, name string) (runtime.Object, error) {
-	if len(kind.Version) == 0 {
-		kind.Version = runtime.APIVersionInternal
+func (t *tracker) List(gvk unversioned.GroupVersionKind, ns string) (runtime.Object, error) {
+	// Heuristic for list kind: original kind + List suffix. Might
+	// not always be true but this tracker has a pretty limited
+	// understanding of the actual API model.
+	listGVK := gvk
+	listGVK.Kind = listGVK.Kind + "List"
+
+	list, err := t.scheme.New(listGVK)
+	if err != nil {
+		return nil, err
+	}
+
+	if !meta.IsListType(list) {
+		return nil, fmt.Errorf("%q is not a list type", listGVK.Kind)
 	}
-	empty, err := o.scheme.New(kind)
-	nilValue := reflect.Zero(reflect.TypeOf(empty)).Interface().(runtime.Object)
 
-	arr, ok := o.types[kind.Kind]
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+
+	objs, ok := t.objects[gvk]
 	if !ok {
-		if strings.HasSuffix(kind.Kind, "List") {
-			itemKind := kind.Kind[:len(kind.Kind)-4]
-			arr, ok := o.types[itemKind]
-			if !ok {
-				return empty, nil
-			}
-			out, err := o.scheme.New(kind)
-			if err != nil {
-				return nilValue, err
-			}
-			if err := meta.SetList(out, arr); err != nil {
-				return nilValue, err
-			}
-			if out, err = o.scheme.Copy(out); err != nil {
-				return nilValue, err
-			}
-			return out, nil
-		}
-		return nilValue, errors.NewNotFound(unversioned.GroupResource{Group: kind.Group, Resource: kind.Kind}, name)
+		return list, nil
 	}
 
-	index := o.last[kind.Kind]
-	if index >= len(arr) {
-		index = len(arr) - 1
+	matchingObjs, err := filterByNamespaceAndName(objs, ns, "")
+	if err != nil {
+		return nil, err
 	}
-	if index < 0 {
-		return nilValue, errors.NewNotFound(unversioned.GroupResource{Group: kind.Group, Resource: kind.Kind}, name)
+	if len(matchingObjs) == 0 {
+		return list, nil
 	}
-	out, err := o.scheme.Copy(arr[index])
+	if err := meta.SetList(list, matchingObjs); err != nil {
+		return nil, err
+	}
+	if list, err = t.scheme.Copy(list); err != nil {
+		return nil, err
+	}
+	return list, nil
+}
+
+func (t *tracker) Get(gvk unversioned.GroupVersionKind, ns, name string) (runtime.Object, error) {
+	obj, err := t.scheme.New(gvk)
```

Comment:
> Not sure how to use scheme to determine whether object is namespaced. ObjectScheme implements ObjectCreater, ObjectCopier and ObjectTyper, none of them seem to have any metadata about the object that we can use to check whether it's namespaced.

Oh, it moved.  `pkg/apimachiner/registered.Group` will give you `GroupMeta` which has a `RESTMapper`, which can give you the `RESTMapping`, which has the `RESTScope`.